### PR TITLE
Allow ETH <> WETH swaps.

### LINKED ISSUE #1168: ETH <> WETH Swaps Are Misbehaving
### Discord Discussion Link

_No response_

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Trying to Swap ETH for WETH

### What did not work?

When trying to swap ETH for WETH - Tally asks the user to "Approve" ETH - which makes no sense since ETH is not an ERC-20 token and should not require approval.  If the user tries to confirm the approval the application hangs.

image

Digging into the network request to 0x in the above example the following payload is sent

```
sellToken: ETH
buyToken: 0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2
gasPrice: 91920471125
slippagePercentage: 0.01
sellAmount: 10000000000000000
affiliateAddress: 0x99b36fDbC582D113aF36A21EBa06BFEAb7b9bE12
feeRecipient: 0x99b36fDbC582D113aF36A21EBa06BFEAb7b9bE12
buyTokenPercentageFee: 0.005
takerAddress: 0x47C7e1eF28F71dd7B5B56ECE1E6fFfcCB9068415
```

with a response of

```
allowanceTarget: "0x0000000000000000000000000000000000000000"
buyAmount: "10000000000000000"
buyTokenAddress: "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"
buyTokenToEthRate: "1"
chainId: 1
estimatedGas: "46000"
estimatedPriceImpact: "0"
gas: "46000"
gasPrice: "91920471125"
minimumProtocolFee: "0"
price: "1"
protocolFee: "0"
sellAmount: "10000000000000000"
sellTokenAddress: "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"
sellTokenToEthRate: "1"
sources: []
value: "10000000000000000"
```

[(click here to see the response in-browser)](https://api.0x.org/swap/v1/price?sellToken=ETH&buyToken=0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2&gasPrice=91920471125&slippagePercentage=0.01&sellAmount=10000000000000000&affiliateAddress=0x99b36fDbC582D113aF36A21EBa06BFEAb7b9bE12&feeRecipient=0x99b36fDbC582D113aF36A21EBa06BFEAb7b9bE12&buyTokenPercentageFee=0.005&takerAddress=0x47C7e1eF28F71dd7B5B56ECE1E6fFfcCB9068415)

There is one things-of-note about the response that suggest that this may be an underlying issue with 0x.

1. The `sellTokenAddress` is not what we would expect to get back from sending a request with a `sellToken` of ETH (we expect `0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee` and instead see that the `sellTokenAddress` is equal to the `buyTokenAddress`

### Version

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by 0xDaedalus**
We need to add a special exception to get ETH <> WETH swaps working because of an issue in the 0x api.  This is a temporary solution and details are commented in code.

Closes #1168 

### To Test
- [ ] Swap ETH into WETH

:Note: It does not seem that it is possible to use fees when wrapping ETH ([sample transaction](https://etherscan.io/tx/0x5a0cd8cf84ab1560f1121b33e2fd7ca703e44fe224c03e95ae30307685b3436d)).

### CODE REVIEW

File: background/redux-slices/0x-swap.ts [hunk 1]
```diff
@@ -249,7 +253,15 @@ export const fetchSwapPrice = createBackgroundAsyncThunk(
...
-    if (quote.sellTokenAddress !== ZEROX_ETH_SIGNIFIER) {
+    if (
+      quote.sellTokenAddress !== ZEROX_ETH_SIGNIFIER &&
+      // When attempting to wrap ETH into WETH or vice-versa - the 0x api currently erroneously returns the requested quote's
+      // sellTokenAddress to the the buyTokenAddress (WETH's contract address).  We have been told that this will be
+      // addressed in a future update to the api (https://github.com/0xProject/0x-api/pull/850) but until that PR
+      // lands we will need to add the below escape hatch to let us slap ETH <> WETH in-app.
+      // @TODO 5/1 Check if we still need the below code.
+      !is0xWETHWrap(quote.buyTokenAddress, quote.sellTokenAddress)
+    ) {
```
Comment (Shadowfiend) [inline background/redux-slices/0x-swap.ts:264]: I still think the note in https://github.com/tallycash/extension/issues/1168#issuecomment-1069437163 is the “most correct” way to handle this—it's both not a special case and happens to match what the API does currently. It also seems more future-proof as a test.

Comfortable merging this first if it feels urgent though.

Comment (0xDaedalus) [reply background/redux-slices/0x-swap.ts]: wow idk how I missed that 😅

**Comment by 0xDaedalus** [comment]
Perhaps a local issue?  Working fine for me:

https://user-images.githubusercontent.com/94649004/160694321-c71b8b53-63c9-4d97-933d-a231c1b4adab.mov

**Review by greg-nagy — APPROVED** [review]
it works now \o/

(I had some really bad mojo going on with my local setup)

**Pull request merged**